ENH: Make 'low' optional in randint

gfyoung opened the pull request:
Added functionality to accept a combination of `low` and `high` parameter values such as
`(None, None)` or `(None, <number>)`. The `(None, None)` option is different from `np.empty` because it doesn't fill an array with just `Py_None` values.  Such options are useful for example if you want to generate numbers for various given `dtype` arguments but don't want to have to keep passing in `low` every time.

The resulting behavior is then defined as follows:

1) `low == None` and `high == None`

Numbers are generated over the range `[lowbnd, highbnd)`, where `lowbnd = np.iinfo(dtype).min`, and `highbnd = np.iinfo(dtype).max`, where `dtype` is the provided integral type.

2) `low != None` and `high == None`

If `low >= 0`, numbers are <b>still</b> generated over the range `[0, low)`, but if `low` < 0, numbers
are generated over the range `[low, highbnd)`, where `highbnd` is defined as above.

3) `low == None` and `high != None`

Numbers are generated over the range `[lowbnd, high)`, where `lowbnd` is defined as above.

Commit:
ENH: Made low optional in randint

Added functionality to accept a combination
of 'low' and 'high' parameter values such as
(None, None) or (None, <number>). The (None,
None) option is different from np.empty because
it doesn't fill an array with 'Py_None' values.

gfyoung commented:
Can someone take a look at this?

madphysicist commented on benchmarks/benchmarks/bench_random.py:
```diff
@@ -34,11 +34,11 @@ class Randint(Benchmark):
     def time_randint_fast(self):
         """Compare to uint32 below"""
-        np.random.randint(0, 2**30, size=10**5)
+        np.random.randint(low=0, high=2**30, size=10**5)
```
I think you should keep the invocation as it was before to ensure backwards compatibility. The same goes for all the other benchmarks and tests you modified in a similar way. If necessary, add a couple more to use the new functionality instead of changing the old ones.

gfyoung replied:
I'm not sure I follow what backwards compatibility you are trying to enforce here.  I see no reason why tests should continue to use a "dated" API.

madphysicist replied:
The API is not dated since I am going to guess that most existing code calls this function as `np.random.randint(somenumber)` and `np.random.randint(somenumber, someothernumber)`. Removing that calling convention from so many of the tests is just asking for trouble later down the line.

gfyoung replied:
The API would become dated if this PR is merged because `low` no longer is a positional argument.  While it may be worth adding <b>a test</b> to ensure the non-keyword argument call still works, I don't think it should persist in the code-base with this change.

juliantaylor replied:
changing non-test code is imo ok, but when changing test code you need to make sure that no coverage of the old still supported way of calling it is lost. It is not obvious if that has been done.
Please add new test cases for the old way of calling the function if you have removed them.

I may have overlooked it but I also see now tests for the new `None` cases

gfyoung replied:
@juliantaylor : The tests are <a href="https://github.com/gfyoung/numpy/blob/randint_no_low/numpy/random/tests/test_random.py#L211">here</a>

ahaldane replied:
I'm also not very enthusuastic about changing all the old unit tests to use the new keywords.

gfyoung replied:
IMHO unit tests should always use the most up-to-date versions of the API for clarity and consistency.  That is why I put one test in this PR for backwards compatibility in which I call the function using the "old" way.

ahaldane replied:
I think guaranteeing that old functionality hasn't broken is more important than checking that new functionality works. I favor virtually never changing old unit tests, unless there is an intentional backwards incompatible change.

I would be OK with you leaving the old unit tests alone, but duplicating some of them and adding keywords in the duplicates, if your new unit tests below don't cover those cases.

gfyoung replied:
I specifically wrote a test <a href="https://github.com/numpy/numpy/pull/7151/files#diff-061a4d644573e126c1d87ab55ad1f5e5R236">here</a> to address that concern.  Maybe it should be expanded, but I don't follow why you would want to have all of your other tests testing the old functionality.

madphysicist commented on numpy/add_newdocs.py:
```diff
@@ -3680,7 +3680,7 @@ def luf(lamdaexpr, *args, **kwargs):
     --------
-    >>> x = np.random.randint(9, size=(3, 3))
+    >>> x = np.random.randint(low=9, size=(3, 3))
     >>> x
```
Keep the example as it was. Add an example showing the function called with no args now that that's possible.

gfyoung replied:
I'm not sure your comment quite applies here because this isn't even documentation for `np.randint` where you are commenting.

madphysicist replied:
Agreed. Removed.

madphysicist commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1159,22 +1159,29 @@ cdef class RandomState:
-            ``high=None``, in which case this parameter is the *highest* such
-            integer).
+        low : int, optional
+            If provided, the lowest (signed) integer to be drawn from the
```
Does it really have to be signed? Same for `high`?

gfyoung replied:
No, it does not.  That's why I think "signed" is placed in parenthesis in the first place : to imply that that may not always be the case.

ahaldane replied:
Maybe you can just leave '(signed)' out? I think the meaning is still clear enough.

gfyoung replied:
Yeah, probably should.  I don't see it having too much of any impact on the meaning.

madphysicist commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1217,15 +1224,25 @@ cdef class RandomState:
-
         key = np.dtype(dtype).name
         if not key in _randint_type:
             raise TypeError('Unsupported dtype "%s" for randint' % key)
```
Should this syntax be updated to Python3? using `str.format` should not break in supported versions of Python2.

gfyoung replied:
I see no reason why it shouldn't be.  Will change.

madphysicist commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1732,17 +1749,17 @@ cdef class RandomState:
         if high is None:
             warnings.warn(("This function is deprecated. Please call "
-                           "randint(1, {low} + 1) instead".format(low=low)),
+                           "randint(low=1, high={low} + 1) instead".format(low=low)),
```
The old version was fine. I think you should revert this line.

gfyoung replied:
I disagree : the inclusion of `low=` and `high=` not only reflects the changed API but also encourages what IMO is good practice in clearly specifying keyword arguments, especially since there are no more positional arguments.

ahaldane replied:
I also like the original better. Even after adding the keyword arguments, I think most numpy users are still going to use the positional version. Plus its more terse in this case.

gfyoung replied:
Well the deprecation _technically_ has not been released in an official version yet, and I also see no reason why people shouldn't be encouraged to use a better style when passing in arguments to essentially an all `kwargs` function.

madphysicist commented on numpy/random/tests/test_random.py:
```diff
@@ -13,19 +13,19 @@
 class TestSeed(TestCase):
     def test_scalar(self):
         s = np.random.RandomState(0)
-        assert_equal(s.randint(1000), 684)
```
Leave the tests the way they were, and add new ones for the new functionality. One for each of the cases

  - no args
  - low=None, high=None
  - low>0, high=None
  - low=0, high=None
  - low<0, high=None
  - low=None, high < 0
  - low=None, high=0
  - low=None, high > 0

gfyoung replied:
See my response above to your comment about backwards compatibility.

madphysicist commented on numpy/random/tests/test_random.py:
```diff
@@ -147,31 +147,31 @@ def test_bounds_checking(self):
             lbnd = 0 if dt is np.bool else np.iinfo(dt).min
             ubnd = 2 if dt is np.bool else np.iinfo(dt).max + 1
             tgt = ubnd - 1
-            assert_equal(self.rfunc(tgt, tgt + 1, size=1000, dtype=dt), tgt)
```
Instead of `tgt + 1`, why not just use `ubnd`?

madphysicist commented on numpy/random/tests/test_random.py:
```diff
@@ -195,19 +195,36 @@ def test_repeatability(self):
         res = hashlib.md5(val).hexdigest()
         assert_(tgt[np.dtype(np.bool).name] == res)
 
+    def test_varying_bounds_args(self):
```
It might be worth adding a case for `low=None, high < 0` just for completeness.

gfyoung replied:
Since I don't have the special behavior like I do for `low != None, high==None`, I see no reason to do this.

madphysicist replied:
Hence "for completeness". The test would ensure that it stays "not a special case".

gfyoung replied:
Hmm...I would think it is the other way around: that you ensure that a special case remains in place.

seberg replied:
I would say the rule of thumb is: one test more won't hurt :)

madphysicist commented:
> 2) `low != None` and `high == None`
> 
> If `low >= 0`, numbers are still generated over the range `[0, low)`, but if `low < 0`, numbers
> are generated over the range `[low, highbnd)`, where `highbnd` is defined as above.

One other thing that bothers me that case #2 has non-intuitive behavior (the old behavior) compared to how the rest of the function now works. At the very least there should be a very emphatic note in the docs saying that the previous behavior is a special case for the new function.

Alternatively, it may be better to introduce a wrapper for `randint` that has completely consistent behavior for all the argument combinations. That is basically what your code is doing anyway. Instead of munging the parameters directly in `randint`, why not keep the original function as-is and have a consistent wrapper that knows what to pass in to it?

gfyoung commented:
I think I will go for the documentation emphasis on backwards incompatibility.  I think the wrapper would add unnecessary complexity to already complicated function.

gfyoung commented:
1) Added test for old non-keyword behavior
2) Expanded documentation and examples for `randint`
3) Added `low = None` and `high < 0` test for completeness

Travis and Appveyor are both happy, so should be good to merge then.

gfyoung commented:
1) Added test for old non-keyword behavior
2) Expanded documentation and examples for `randint`
3) Added `low = None` and `high < 0` test for completeness

Travis and Appveyor are both happy, so should be good to merge then.

gfyoung commented:
1) Added test for old non-keyword behavior
2) Expanded documentation and examples for `randint`
3) Added `low = None` and `high < 0` test for completeness

Travis and Appveyor are both happy, so should be good to merge then.

gfyoung commented:
1) Added test for old non-keyword behavior
2) Expanded documentation and examples for `randint`
3) Added `low = None` and `high < 0` test for completeness

Travis and Appveyor are both happy, so should be good to merge then.

gfyoung commented:
1) Added test for old non-keyword behavior
2) Expanded documentation and examples for `randint`
3) Added `low = None` and `high < 0` test for completeness

Travis and Appveyor are both happy, so should be good to merge then.

gfyoung commented:
1) Added test for old non-keyword behavior
2) Expanded documentation and examples for `randint`
3) Added `low = None` and `high < 0` test for completeness

Travis and Appveyor are both happy, so should be good to merge then.

gfyoung commented:
Can someone with merging capabilities take a look at this?  I think it is ready to go.

ahaldane commented:
This is a good idea, and the behavior looks right.

I agree with some of the comments above though, where I replied.

ahaldane commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1159,22 +1159,30 @@ cdef class RandomState:
+        where `lowbnd` is defined as previously. If only `high` is None,
+        then if `low` > 0, results are from [0, `low`). This behaviour is
+        intended to maintain **backwards compatibility**. Otherwise, results
+        are from [`low`, `highbnd`), where `highbnd` is defined as previously.
```
I agree with @madphysicist here that instead of saying "This behaviour is intended to maintain **backwards compatibility**", I would prefer to say something "This is so that randint(n) will give a random integer from 0 to n".

I think most/many uses of randint are still going to use positional arguments, even after this PR.

gfyoung replied:
The latter explanation is confusing IMO: it seems very out of place when you compare it to the case when `low=None` and `high != None`.  Backwards compatibility is the sole reason for me leaving the behavior as it is.  Otherwise, I would have changed it.

Probably many will still use `randint` with position arguments, but nevertheless people should be encouraged to adapt to the new API since it technically isn't breaking.  I'm not sure I follow your reasoning as to why function calls in the code-base should continue to follow positional arguments when using `=` is a lot cleaner and clearer just because of ingrained user habits.

madphysicist commented:
Just a question: have you bounced this off the mailing list? If so, would you mind posting the link here for reference? While I am personally not a huge fan of this API change, I do not think that should stop the change from happening if the consensus is that people would like to see it done.

gfyoung commented:
@madphysicist : There is one now.

madphysicist commented:
I appreciate that. I think that if you get a lot of positive response, it will help convince the devs to merge your changes.

For future reference, here is the archive link: https://mail.scipy.org/pipermail/numpy-discussion/2016-February/075018.html

gfyoung commented:
@madphysicist : Well certainly a positive response from the general mailing list couldn't hurt! :)

gfyoung commented:
Alas, after posting to the email list, there didn't seem to be much support for this, so closing.
